src/protocol/coap.rs

```diff
@@ -0,0 +1,130 @@
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
+pub enum Method {
+    Get,
+    Put,
+    Post,
+    Delete,
+    Patch,
+    #[serde(rename = "iPATCH")]
+    Ipatch,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, Default)]
+pub struct BlockWiseTransferParameters {
+    #[serde(rename = "cov:block2SZX")]
+    block_2szx: Option<u16>,
+    #[serde(rename = "cov:block1SZX")]
+    block_1szx: Option<u16>,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, Default)]
+// #[serde(rename_all = "camelCase")]
+pub struct Form {
+    #[serde(rename = "cov:method")]
+    method: Option<Method>,
+    #[serde(rename = "cov:blockwise")]
+    blockwise: Option<BlockWiseTransferParameters>,
+    #[serde(rename = "cov:qblockwise")]
+    qblockwise: Option<BlockWiseTransferParameters>,
+    #[serde(rename = "cov:hopLimit")]
+    hop_limit: Option<u8>,
+}
+
+use crate::extend::ExtendableThing;
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct CoapProtocol {}
+
+impl ExtendableThing for CoapProtocol {
+    type InteractionAffordance = ();
+    type PropertyAffordance = ();
+    type ActionAffordance = ();
+    type EventAffordance = ();
+    type Form = Form;
+    type ExpectedResponse = ();
+    type DataSchema = ();
+    type ObjectSchema = ();
+    type ArraySchema = ();
+}
+
+#[cfg(test)]
+mod test {
+    use super::CoapProtocol;
+    use crate::hlist::{Cons, Nil};
+    fn deserialize_form(s: &str) {
+        let f: super::Form = serde_json::from_str(s).unwrap();
+
+        dbg!(&f);
```

Comment:
I would remove these `dbg`. Moreover, I don't like this approach: instead of testing if the deserialization produces a specific structure, you are somehow testing the extension capabilities of `Form`, which is totally out of scope.

Instead, test if your structure deserializes correctly and produce a specific structure, without taking into account `Form`. Keep in mind that there using your approach you are barely testing serde, not your implementation.